cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -919,25 +916,30 @@ func (csr *ClusterStateRegistry) GetIncorrectNodeGroupSize(nodeGroupName string)
 }
 
 // GetUpcomingNodes returns how many new nodes will be added shortly to the node groups or should become ready soon.
-// The function may overestimate the number of nodes.
-func (csr *ClusterStateRegistry) GetUpcomingNodes() map[string]int {
+// The function may overestimate the number of nodes. The second return value contains the names of upcoming nodes
+// that are already registered in the cluster.
+func (csr *ClusterStateRegistry) GetUpcomingNodes() (upcomingCounts map[string]int, registeredNodeNames map[string][]string) {
 	csr.Lock()
 	defer csr.Unlock()
 
-	result := make(map[string]int)
+	upcomingCounts = map[string]int{}
+	registeredNodeNames = map[string][]string{}
 	for _, nodeGroup := range csr.cloudProvider.NodeGroups() {
 		id := nodeGroup.Id()
 		readiness := csr.perNodeGroupReadiness[id]
 		ar := csr.acceptableRanges[id]
 		// newNodes is the number of nodes that
-		newNodes := ar.CurrentTarget - (readiness.Ready + readiness.Unready + readiness.LongUnregistered)
+		newNodes := ar.CurrentTarget - (len(readiness.Ready) + len(readiness.Unready) + len(readiness.LongUnregistered))
 		if newNodes <= 0 {
 			// Negative value is unlikely but theoretically possible.
 			continue
 		}
-		result[id] = newNodes
+		upcomingCounts[id] = newNodes
+		// newNodes should be roughly equal to readiness.NotStarted + readiness.Unregistered. NotStarted are nodes that are
+		// already registered in the cluster but not yet ready, Unregistered are nodes that aren't registered in the cluster yet.
+		registeredNodeNames[id] = readiness.NotStarted
```

Comment:
nit: I would personally return all known instance names (ie. including unregistered) and later filter out from snapshot the ones that are there. Or maybe return both registered and unregistered lists as separate values. 
The interface where we return one, but not the other seems strange to me. But, of course, we don't need unregistered node names for anything in this codepath and my aesthetic feelings don't feel like strong enough justification, so feel free to ignore this comment.

Reply:
I agree it's not ideal, but I'm afraid returning them both in the same object could be confusing, since they'd have a different format depending on whether they're registered or not (unregistered nodes get the instance id as a placeholder name, which is probably actually an URL with more info, not just the name) - and this wouldn't be obvious to the caller. Returning both in separate objects would solve that and would be cleaner, I just didn't want to do that just to discard at the call site. We can always add the second object if needed in the future.